src/runtime_src/core/common/dev_factory.h

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Advanced Micro Devices, Inc. All rights reserved.
+#ifndef _XCL_DEV_FACTORY_H_
+#define _XCL_DEV_FACTORY_H_
+
+#include "device.h"
+#include <vector>
+
+namespace xrt_core {
+
+// One function on FPGA or AIE device
+class dev
+{
+public:
+  bool m_is_mgmt = false;
+  bool m_is_ready = true; 
+  dev(bool isuser) {  
+      m_is_mgmt = !isuser;  
+  }
+  
+  virtual ~dev() {}
+ 
+  // Hand out a "device" instance that is specific to this type of device.
+  // Caller will use this device to access device specific implementation of ishim.
+  virtual std::shared_ptr<device>
+  create_device(device::handle_type handle, device::id_type id) const = 0;
+
+  // Hand out an opaque "shim" handle that is specific to this type of device.
+  // On legacy Alveo device, this handle can be used to lookup a device instance and
+  // make xcl HAL API calls.
+  // On new platforms, this handle can only be used to look up a device. HAL API calls
+  // through it are not supported any more.
+  virtual device::handle_type
+  create_shim(device::id_type id) const = 0;
+
+  //get BDF 
+  virtual std::tuple<uint16_t, uint16_t, uint16_t, uint16_t>
+  get_bdf_info() const
+  {
+    return std::make_tuple(uint16_t(0), uint16_t(0), uint16_t(0), uint16_t(0));
+  }
+};
+
+
+size_t
+get_dev_ready(bool user = true);
+
+size_t
+get_dev_total(bool user = true);
+
+std::shared_ptr<dev>
+get_dev(unsigned index, bool user = true);
+
+
+/**
+ * Adding device_factory instance to the global list. Should only be called from global object of
+ * built-in drivers through dlopen.
+ * For now, once added, it cannot be removed until the list itself is out of scope.
+ */
+XRT_CORE_COMMON_EXPORT
+void
+register_device_list(const std::vector<std::shared_ptr<xrt_core::dev>>& devlist);
```

Comment:
I will need an explanation of the object ownership to understand this.   For example, I would think the shim library would create device factories upon load and pass ownership of these to XRT.   I am also questioning why this is not just a "register one factory" at a time interface?

Reply:
yes soren, we are registering list of device_factory objects during loading of each driver( dlopen of .so) , we can make it one device registration instead of list.